### PR TITLE
Implement vote transaction

### PR DESCRIPTION
**Pull request by MSNTCS**

### CODE REVIEW

File: voting/src/voting.rs [hunk 1]
```diff
@@ -198,6 +198,27 @@ impl TxOwner for Context {
...
+                let choice = tx.action.choice;
+                let voter_signature = tx.action.voter_signature;
+
+                if choice != FAVOR || choice != AGAINST || choice != ABSENTION {
+                    return Err(ExecuteError::InvalidChoice).map_err(|_| todo_fixthis)
+                }
+                let mut vote_paper: VotePaper =
+                    serde_cbor::from_slice(&self.storage().get(vote_paper_id.0.as_ref()).unwrap())
+                        .map_err(|_| ExecuteError::VotePaperNotFound)
+                        .map_err(|_| todo_fixthis)?;
```
Comment (majecty) [inline voting/src/voting.rs:213]: Please take care when you use `unwrap`.

```suggestion
                    serde_cbor::from_slice(&self.storage().get(vote_paper_id.0.as_ref()).map_err(|_| ExecuteError::VotePaperNotFound)?)
                        .unwrap()
                        .map_err(|_| todo_fixthis)?;
```

File: voting/src/voting.rs [hunk 2]
```diff
@@ -51,8 +55,17 @@ pub struct TxCreateVotePaper {
     voter_public_key: Public,
 }
 
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TxVote {
+    vote_paper_id: H256,
+    choice: String,
```
[the remark this answers is not in the document]
Comment (MSNTCS) [reply voting/src/voting.rs:61]: @majecty Do you mean choice?

[the remark this answers is not in the document]
Comment (majecty) [reply voting/src/voting.rs:61]: Yes using an enum for the choice field.

File: voting/src/voting.rs [hunk 3]
```diff
@@ -101,6 +114,36 @@ impl Context {
...
+                let vote = Vote::new(vote_paper_id, choice, voter_signature);
+                let vote_id = vote.clone().vote_id;
+                vote_paper.set_used_in(vote_id.clone()).map_err(|_| ExecuteError::UsedVotePaper)?;
+                self.storage_mut().set(vote_id.as_ref(), serde_cbor::to_vec(&vote).unwrap());
+
+                let meeting_id = vote_paper.get_general_meeting_id();
+
+                let box_key = crate::generate_voting_box_key(&meeting_id);
+                let mut vote_box = self.meeting_mut().get_vote_box(box_key.as_slice()).unwrap();
+                vote_box.drop_in_box(vote.clone().vote_id);
```
[the remark this answers is not in the document]
Comment (MSNTCS) [reply voting/src/voting.rs:143]: Yes, we need to have a service like this.

File: voting/src/general_meeting.rs [hunk 4]
```diff
@@ -225,6 +226,17 @@ impl GeneralMeetingManager for Context {
...
+    fn set_vote_box(&mut self, meeting_id: &GeneralMeetingId, vote_box: &VoteBox) -> Result<(), Error> {
+        let key = crate::generate_voting_box_key(meeting_id);
+
+        if !self.storage().has(key.as_slice()) {
+            return Err(Error::VoteBoxNotFound)
+        }
+
+        self.storage_mut().set(key.as_slice(), serde_cbor::to_vec(&vote_box).unwrap());
+        Ok(())
+    }
```
Comment (majecty) [inline voting/src/general_meeting.rs:239]: Returning error if votebox is not exist when setting seems strange.
How about `update_vote_box`?

When I read the `set`, I feel like it will success in the both case there is the previous state or empty.

File: voting/src/voting.rs [hunk 5]
```diff
@@ -101,6 +114,47 @@ impl Context {
...
                 Ok(vote_paper.vote_paper_id)
             }
+            VOTE_TX_TYPE => {
+                let tx: VoteTransaction =
+                    serde_cbor::from_slice(&transaction.body()).map_err(|_| ExecuteError::InvalidFormat)?;
+                let vote_paper_id = tx.action.vote_paper_id;
+                let choice = tx.action.choice;
+                let voter_signature = tx.action.voter_signature;
+
+                if choice != VoteChoice::Favor || choice != VoteChoice::Against || choice != VoteChoice::Absention {
```
Comment (majecty) [inline voting/src/voting.rs:124]:
```suggestion
                if choice != VoteChoice::Favor && choice != VoteChoice::Against && choice != VoteChoice::Absention {
```

File: voting/src/voting.rs [hunk 6]
```diff
@@ -155,6 +209,36 @@ impl TxOwner for Context {
                 }
                 Ok(())
             }
+            VOTE_TX_TYPE => {
+                let tx: VoteTransaction = serde_cbor::from_slice(&transaction.body()).map_err(|_| todo_fixthis)?;
+                let vote_paper_id = tx.action.vote_paper_id;
+                let choice = tx.action.choice;
+                let voter_signature = tx.action.voter_signature;
+
+                if choice != VoteChoice::Favor || choice != VoteChoice::Against || choice != VoteChoice::Absention {
```
Comment (majecty) [inline voting/src/voting.rs:218]:
```suggestion
                if choice != VoteChoice::Favor && choice != VoteChoice::Against && choice != VoteChoice::Absention {
```

File: voting/src/voting.rs [hunk 7]
```diff
@@ -101,6 +114,45 @@ impl Context {
...
+                let tx: VoteTransaction =
+                    serde_cbor::from_slice(&transaction.body()).map_err(|_| ExecuteError::InvalidFormat)?;
+                let vote_paper_id = tx.action.vote_paper_id;
+                let choice = tx.action.choice;
+                let voter_signature = tx.action.voter_signature;
+
+                if choice != VoteChoice::Favor && choice != VoteChoice::Against && choice != VoteChoice::Absention {
+                    return Err(ExecuteError::InvalidChoice)
+                }
+
```
Comment (majecty) [inline voting/src/voting.rs:127]: Let's remove it.
We are using `VoteChoice` enum. If the choice is not favor, against, and abesention, it will cause deserialization error.

File: voting/src/voting.rs [hunk 8]
```diff
@@ -101,6 +114,45 @@ impl Context {
...
+                if !self.storage().has(vote_paper_id.0.as_ref()) {
```
Comment (majecty) [inline voting/src/voting.rs:128]:
```suggestion
                if !self.storage().has(vote_paper_id.as_ref()) {
```

File: voting/src/voting.rs [hunk 9]
```diff
@@ -151,6 +151,8 @@ impl Context {
                 let mut vote_box = self.meeting_mut().get_vote_box(box_key.as_slice()).unwrap();
                 vote_box.drop_in_box(vote.vote_id.clone());
 
+                self.meeting_mut().update_vote_box(&meeting_id, &vote_box);
```
Comment (majecty) [inline voting/src/voting.rs:154]: Please do not ignore returned Result type.
```suggestion
                self.meeting_mut().update_vote_box(&meeting_id, &vote_box).unwrap("We get the vote box before");
```

File: voting/src/general_meeting.rs [hunk 10]
```diff
@@ -225,6 +226,17 @@ impl GeneralMeetingManager for Context {
         };
         Ok(vote_box)
     }
+
+    fn update_vote_box(&mut self, meeting_id: &GeneralMeetingId, vote_box: &VoteBox) -> Result<(), Error> {
```
Comment (majecty) [inline voting/src/general_meeting.rs:230]: `get_vote_box` receives `&[u8]` as key. `update_vote_box` receives `meeting_id: &GeneralMeetingId` as key.
Please use a consistent type to find a vote box.

IMO, `meeting_id: &GeneralMeetingId` is better than `&[u8]`. How about making `get_vote_box` to receive `&GeneralMeetingId` as argument instead of `key: &[u8]`?

File: voting/src/voting.rs [hunk 11]
```diff
@@ -216,10 +218,19 @@ impl TxOwner for Context {
...
+                }
+
+                let mut vote_paper: VotePaper = {
+                    let bytes = &self
+                        .storage()
+                        .get(vote_paper_id.0.as_ref())
+                        .expect("Previously we checked the existence of the vote paper");
+                    serde_cbor::from_slice(bytes).expect("Vote paper is serialized by this code")
+                };
+
```
Comment (majecty) [inline voting/src/voting.rs]: This looks not related to "Add a missing function to update vote box changes" commit.

File: voting/src/voting.rs [hunk 12]
```diff
@@ -101,6 +114,45 @@ impl Context {
...
+                    return Err(ExecuteError::InvalidChoice)
+                }
+
+                if !self.storage().has(vote_paper_id.0.as_ref()) {
+                    return Err(ExecuteError::VotePaperNotFound)
+                }
+                let mut vote_paper: VotePaper = {
+                    let bytes = self
+                        .storage()
+                        .get(vote_paper_id.0.as_ref())
```
Comment (majecty) [inline voting/src/voting.rs:134]:
```suggestion
                        .get(vote_paper_id.as_ref())
```

File: voting/src/voting.rs [hunk 13]
```diff
@@ -101,6 +114,45 @@ impl Context {
...
+                let vote_id = vote.vote_id.clone();
+                vote_paper.set_used_in(vote_id.clone()).map_err(|_| ExecuteError::UsedVotePaper)?;
+                self.storage_mut().set(vote_id.as_ref(), serde_cbor::to_vec(&vote).unwrap());
+
+                let meeting_id = vote_paper.get_general_meeting_id();
+
+                let box_key = crate::generate_voting_box_key(&meeting_id);
+                let mut vote_box = self.meeting_mut().get_vote_box(box_key.as_slice()).unwrap();
+                vote_box.drop_in_box(vote.vote_id.clone());
+
```
Comment (majecty) [inline voting/src/voting.rs:153]: How about moving this code to "Add a missing function to update vote box changes" commit?

File: voting/src/voting.rs [hunk 14]
```diff
@@ -217,6 +217,57 @@ impl TxOwner for Context {
...
+        if !self.storage().has(vote_id.as_ref()) {
+            return Err(Error::VoteNotFound)
+        }
+        let vote: Vote = {
+            let bytes = &self.storage().get(vote_id.as_ref()).expect("Previously we checked the existence of the vote");
+            serde_cbor::from_slice(bytes).expect("Vote is serialized by this code")
+        };
+        let vote_paper_id = vote.corresponding_paper_id;
+
+        if !self.storage().has(vote_paper_id.0.as_ref()) {
```
Comment (majecty) [inline voting/src/voting.rs:288]:
```suggestion
        if !self.storage().has(vote_paper_id.as_ref()) {
```

File: voting/src/voting.rs [hunk 15]
```diff
@@ -217,6 +217,57 @@ impl TxOwner for Context {
...
+            return Err(Error::CorrespondingVotePaperNotFound)
+        }
+        let vote_paper: VotePaper = {
+            let bytes = &self
+                .storage()
+                .get(vote_paper_id.0.as_ref())
```
Comment (majecty) [inline voting/src/voting.rs:294]:
```suggestion
                .get(vote_paper_id.as_ref())
```

File: voting/src/voting.rs [hunk 16]
```diff
@@ -207,6 +207,27 @@ impl TxOwner for Context {
...
             }
+            VOTE_TX_TYPE => {
+                let tx: VoteTransaction = serde_cbor::from_slice(&transaction.body()).map_err(|_| todo_fixthis)?;
+                let vote_paper_id = tx.action.vote_paper_id;
+                let choice = tx.action.choice;
+                let voter_signature = tx.action.voter_signature;
+
+                if choice != VoteChoice::Favor && choice != VoteChoice::Against && choice != VoteChoice::Absention {
+                    return Err(ExecuteError::InvalidChoice).map_err(|_| todo_fixthis)
+                }
```
Comment (majecty) [inline voting/src/voting.rs:220]: Let's remove this code.

We are using enum for the VoteChoice. If the choice is not one of Favor, Against, and Absention, there will be a deserialization error.

File: voting/src/voting.rs [hunk 17]
```diff
@@ -207,6 +207,27 @@ impl TxOwner for Context {
...
+                let mut vote_paper: VotePaper =
+                    serde_cbor::from_slice(&self.storage().get(vote_paper_id.0.as_ref()).unwrap())
```
Comment (majecty) [inline voting/src/voting.rs]:
```suggestion
                    serde_cbor::from_slice(&self.storage().get(vote_paper_id.as_ref()).unwrap())
```

Comment (majecty) [inline voting/src/voting.rs]:
```suggestion
                    serde_cbor::from_slice(&self.storage().get(vote_paper_id.0.as_ref()).map_err(|_| ExecuteError::VotePaperNotFound))
```

File: voting/src/voting.rs [hunk 18]
```diff
@@ -207,6 +207,27 @@ impl TxOwner for Context {
...
+                        .map_err(|_| ExecuteError::VotePaperNotFound)
```
Comment (majecty) [inline voting/src/voting.rs]:
```suggestion
                        .expect("Vote paper is serialized by this code")
```

File: voting/src/voting.rs [hunk 19]
```diff
@@ -216,10 +218,19 @@ impl TxOwner for Context {
                 if choice != VoteChoice::Favor && choice != VoteChoice::Against && choice != VoteChoice::Absention {
                     return Err(ExecuteError::InvalidChoice).map_err(|_| todo_fixthis)
                 }
-                let mut vote_paper: VotePaper =
-                    serde_cbor::from_slice(&self.storage().get(vote_paper_id.0.as_ref()).unwrap())
-                        .map_err(|_| ExecuteError::VotePaperNotFound)
-                        .map_err(|_| todo_fixthis)?;
+
+                if !self.storage().has(vote_paper_id.0.as_ref()) {
```
Comment (majecty) [inline voting/src/voting.rs:222]:
```suggestion
                if !self.storage().has(vote_paper_id.as_ref()) {
```

File: voting/src/voting.rs (hunk 8 above)
Comment (MSNTCS) [reply voting/src/voting.rs:128]: @majecty  id has the H256 type. `as_ref` is not implemented for it.

Comment (majecty) [reply voting/src/voting.rs:128]: `H256` implements `AsRef<u8>`
https://github.com/paritytech/parity-common/blob/fixed-hash-v0.2.5/fixed-hash/src/hash.rs#L60

File: voting/src/voting.rs [hunk 20]
```diff
@@ -38,6 +38,7 @@ enum ExecuteError {
     VoteAfterTallyingTime,
     InvalidAgendaNumber,
     UsedVotePaper,
+    VotePaperNotFound,
```
Comment (majecty) [inline voting/src/voting.rs:41]: Nice to see a separate commit for this: "Add get_vote_paper function".

File: voting/src/voting.rs [hunk 21]
```diff
@@ -102,6 +113,29 @@ impl Context {
...
             }
+            VOTE_TX_TYPE => {
+                let tx: VoteTransaction =
+                    serde_cbor::from_slice(&transaction.body()).map_err(|_| ExecuteError::InvalidFormat)?;
+                let vote_paper_id = tx.action.vote_paper_id;
+                let choice = tx.action.choice;
+                let voter_signature = tx.action.voter_signature;
+
+                let mut vote_paper =
+                    self.get_vote_paper(&vote_paper_id).map_err(|_| ExecuteError::VotePaperNotFound)?;
```
Comment (majecty) [inline voting/src/voting.rs:124]:
```suggestion
                    self.get_vote_paper(&vote_paper_id)?;
```

**Comment by MSNTCS** [comment]
> There are many points that read or write `Vote` and `VotePaper`.
> How about creating a getter and setter in Context?
> Like `get_vote_paper` and `set_vote_paper`.

I think get

File: voting/src/voting.rs [hunk 22]
```diff
@@ -102,6 +113,29 @@ impl Context {
...
+
+                if !vote_paper.verify_signature(&voter_signature) {
+                    return Err(ExecuteError::InvalidVoterSignature)
+                }
+
+                let vote = Vote::new(vote_paper_id, choice, voter_signature);
+                let vote_id = vote.vote_id.clone();
+                vote_paper.set_used_in(vote_id.clone()).map_err(|_| ExecuteError::UsedVotePaper)?;
```
Comment (majecty) [inline voting/src/voting.rs:132]:
```suggestion
                vote_paper.set_used_in(vote_id.clone())?;
```

File: voting/src/voting.rs [hunk 23]
```diff
@@ -206,6 +206,25 @@ impl TxOwner for Context {
...
             }
+            VOTE_TX_TYPE => {
+                let tx: VoteTransaction = serde_cbor::from_slice(&transaction.body()).map_err(|_| todo_fixthis)?;
+                let vote_paper_id = tx.action.vote_paper_id;
+                let choice = tx.action.choice;
+                let voter_signature = tx.action.voter_signature;
+
+                let mut vote_paper = self
+                    .get_vote_paper(&vote_paper_id)
+                    .map_err(|_| ExecuteError::VotePaperNotFound)
```
Comment (majecty) [inline voting/src/voting.rs:222]:
```suggestion
                    .get_vote_paper(&vote_paper_id)
```

File: voting/src/voting.rs [hunk 24]
```diff
@@ -101,9 +113,53 @@ impl Context {
...
+                let vote_id = vote.vote_id.clone();
+                vote_paper.set_used_in(vote_id.clone()).map_err(|_| ExecuteError::UsedVotePaper)?;
+                self.storage_mut().set(vote_id.as_ref(), serde_cbor::to_vec(&vote).unwrap());
+
+                let meeting_id = vote_paper.get_general_meeting_id();
+
+                let mut vote_box = self.meeting_mut().get_vote_box(&meeting_id).unwrap();
+                vote_box.drop_in_box(vote.vote_id.clone());
+
+                self.meeting_mut().update_vote_box(&meeting_id, &vote_box);
```
Comment (majecty) [inline voting/src/voting.rs:140]: I missed this. Please do not ignore returned Result.

```suggestion
                self.meeting_mut().update_vote_box(&meeting_id, &vote_box).expect("Votebox is created when creating a general meeting.");
```
